syntax = "proto2";

package courses;

message CourseResponse {
  required string id = 1;
  required string name = 2;
  optional string description = 3;
  required string author_id = 4;
}

message CourseCreateRequest {
  required string name = 1;
  optional string description = 2;
  required string author_id = 3;
}

message CourseFindByIdRequest {
  required string id = 1;
}

message CourseFindRequest {
  optional string name = 1;
  optional string author_id = 2;
  optional string user_id = 3;
  required uint64 limit = 4;
  required uint64 offset = 5;
}

message CourseFindResponse {
  repeated CourseResponse results = 1;
  required uint64 total_count = 2;
}

message AccessRequest {
  required string course_id = 1;
  required string user_id = 2;
  required bool access = 3;
}

message AccessResponse {
  required string user_id = 1;
}

message CourseAccessRequest {
  required string course_id = 1;
}

message CourseAccessResponse {
  repeated AccessResponse results = 1;
  required uint64 total_count = 2;
}

service Courses {
  rpc CreateCourse(CourseCreateRequest) returns (CourseResponse);
  rpc FindCourseById(CourseFindByIdRequest) returns (CourseResponse);
  rpc FindCourses(CourseFindRequest) returns (CourseFindResponse);
  rpc ModifyAccess(AccessRequest) returns (AccessResponse);
  rpc GetCourseAccess(CourseAccessRequest) returns (CourseAccessResponse);
}